### PR TITLE
Specify types on get_parent_class()

### PR DESCRIPTION
**Pull request by staabm**
closes https://github.com/phpstan/phpstan/issues/5552

### CODE REVIEW

File: src/Analyser/TypeSpecifier.php [hunk 1]
```diff
@@ -267,6 +267,47 @@ public function specifyTypesInCondition(
...
+				if (
+					!$context->null()
+					&& $exprNode instanceof FuncCall
+					&& $exprNode->name instanceof Name
+					&& strtolower((string) $exprNode->name) === 'get_parent_class'
+					&& isset($exprNode->getArgs()[0])
+					&& $constantType instanceof ConstantStringType
+				) {
+					$argType = $scope->getType($exprNode->getArgs()[0]->value);
+					$objectType = new ObjectType($constantType->getValue());
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php]: the reported phpstan error is interessting..

grafik

shouldn't phpstan know, that `$constantType is a ConstantStringType` and therefore know that `getValue` can only return a string?

Comment (ondrejmirtes) [reply src/Analyser/TypeSpecifier.php]: It should and it works in other places too. My guess is that you broke this functionality in this PR, but it's hard to say just by looking. You need to debug it.

Comment (staabm) [reply src/Analyser/TypeSpecifier.php]: the issue I am experiencing here seems not be triggered by my changes. I think https://github.com/phpstan/phpstan-src/pull/1396 proves that the problem also happens on the latest master without further modifications

**Comment by rvanvelzen** [comment]
Check this strange behavior as well:

- https://phpstan.org/r/889e3c66-972b-4ace-af00-e9beb3280918 (with `dumpType($constantType)`)
- https://phpstan.org/r/1e85e658-425f-44ef-afeb-7243c3df5221 (_without_ `dumpType($constantType)`)

File: src/Analyser/TypeSpecifier.php [hunk 2]
```diff
@@ -326,6 +326,50 @@ public function specifyTypesInCondition(
...
+
+				if (
+					!$context->null()
+					&& $exprNode instanceof FuncCall
+					&& $exprNode->name instanceof Name
+					&& strtolower((string) $exprNode->name) === 'get_parent_class'
+					&& isset($exprNode->getArgs()[0])
+					&& $constantType instanceof ConstantStringType
+					// the is_string() check should not be necessary, since ConstantStringType->getValue() returns 'string'
+					&& is_string($constantType->getValue())
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php]: this works arround the type error:

grafik

which is a separate bug, see https://github.com/phpstan/phpstan-src/pull/1396

Comment (staabm) [reply src/Analyser/TypeSpecifier.php]: in https://github.com/phpstan/phpstan-src/pull/1383/commits/fb01d178d5f5bd0db294fcfb6ef47ca2d2d769ee I factored out the existing logic into 2 new private methods. 

that way we don't run into the https://github.com/phpstan/phpstan-src/pull/1396 bug and don't need the `is_string` based workaround

File: src/Analyser/TypeSpecifier.php [hunk 3]
```diff
@@ -1034,6 +896,215 @@ public function specifyTypesInCondition(
...
+			if ($constantType->getValue() === 'object') {
+				$type = new ObjectWithoutClassType();
+			}
+
+			if ($type !== null) {
+				return $this->create($exprNode->getArgs()[0]->value, $type, $context, false, $scope, $rootExpr);
+			}
+		}
+
+		if (
```
Comment (staabm) [inline src/Analyser/TypeSpecifier.php:1065]: this if block is the only new logic of this PR.

everything else is pre-existing logic which was just factored out into new private methods

File: src/Analyser/TypeSpecifier.php (hunk 2 above)
Comment (staabm) [reply src/Analyser/TypeSpecifier.php]: fyi @rvanvelzen, the refactoring might conflict with your existing https://github.com/phpstan/phpstan-src/pull/1761 PR.

**Comment by staabm** [comment]
@ondrejmirtes if we could land this one it would unblock some of the already opened PRs for more TypeSpecifier goodness.

**Comment by ondrejmirtes** [comment]
Thank you.

**Pull request merged**